keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,556 @@
+* `-v` flag should allow to change verbosity of logging
+
+Any additional configuration needed by loggers should be provided by flags specific to its implementation. Default loggers should provide a basic customization, but it will not be encouraged to avoid feature bloat. Adding new loggers, changing logger behavior or extending its functionality should be done outside of Kubernetes by implementing logr API and registering it in code.
+
+We acknowledge that extension via code is not the best, but should be sufficient for now.
+
+Proposed flag `--logging-format`  values:
+* `legacy` for old logging format
+* `json` for json format
+
+All logs from the `json` logger should go to standard output.  Out of the existing klog command line arguments only ‘-v’ with be supported in future.
+
+### Kubernetes log schema
+
+Introducing structured logging API is not enough to address one of the biggest issues with log creation or analysis, that is inconsistent metadata. Size of Kubernetes codebase could easily result in drifting naming conventions. As part of this proposal we want to propose a mechanism for ensuring consistency of metadata. To achieve this we will introduce a way to define a schema for subset of commonly used fields in kubernetes. This schema should improve reusability of fields and provide discovery mechanism for logging backend vendors.
+
+Introducing this schema will include three things:
+* Schema definition format
+* Set of helper functions to populate this schema (e.g. `WithObject` for referencing k8s object)
+* Validation mechanism ensuring schema is not bypassed by lower level logging API
+
+Exact schema definition is outside the scope of this proposal and will be addressed by a separate KEP when structured logging effort will enter Beta and logs will start to be migrated. At the end we would expect schema to look similar to [Elastic Common Schema](https://github.com/elastic/ecs/tree/master/schemas)
+
+### Migration
+
+Migration to new structured logging API will be divided into 4 stages:
+
+#### Stage 1 / Alpha
+
+This stage needs to be performed in a single release of Kubernetes. When finished for every **existing** log entry produced in legacy output format there should be a corresponding log entry in json format containing similar information possibly with some additional context information attached to it.
+When using the legacy log format after this stage Kubernetes control plane components should produce exactly the same logs as before this stage.
+After this stage format and content of json formatted logs will be still in pre-alpha state and most likely will change after manual review.
+After completing this stage adding **new log entries should require using only the new logr-based logging API**. Double writing using both klog and logr APIs for new log entries should not be needed.
```

Comment:
how do you plan to enforce this?

Reply:
This is no longer needed. For enforcing such changes we previously used static analysis on [AST](https://en.wikipedia.org/wiki/Abstract_syntax_tree). This method currently is used for enforcing review of new Conformance tests and stable metrics.